### ISSUE TITLE
Alerting Screenshot timerange limited to 1 hour

### ISSUE THREAD

**Issue by apollo13** [grafana/grafana#63718]
**What happened**:

After upgrading from 9.3 to 9.4 beta1 our alerting screenshots stopped showing values. After a bit of debugging we realized that the rendering URL now includes `from=now-1h&to=now` by default: #59166 (cc @grobinson-grafana since you might have some deeper insights here). Before it included nothing which as far as I can tell simply used the Panel time (in our case a week).

**What you expected to happen**:

Keep existing images working and display what the panel shows.

**How to reproduce it (as minimally and precisely as possible)**:

Create a timeseries panel that shows data from a week (explicit range override on the panel), alert on it and observe the timerange in the rendered image.

**Anything else we need to know?**:

It would be great if it where possible to configure the timerange per alert or at least take it from the panel.

**Environment**:
- Grafana version: 9.4 beta 1
- Data source type & version:
- OS Grafana is installed on: Docker
- User OS & Browser: headless chrome?
- Grafana plugins:
- Others:

**Comment by grobinson-grafana** [comment]
Hi! :wave: I think what we should do here is use the `For` interval of the alert rule when taking the screenshot, as I suspect most people don't change the default panel time to match their alert rule, resulting in screenshots that are over too long a time period to show when the alert fired.

**Comment by apollo13** [comment]
Hi @grobinson-grafana, thank you for your answer. Do you mean the `For` from the following screenshot?

![Screenshot from 2023-03-03 13-09-41](https://user-images.githubusercontent.com/81547/222717102-01625df4-406f-4dd4-91eb-9966f9920fbb.png)

If yes then I am not sure I agree. Depending on the type of alert for instance we often set this to 0s to alert immediately. 

Wouldn't it make more sense to use the time range from the alert itself (see below)?

![Screenshot from 2023-03-03 13-14-52](https://user-images.githubusercontent.com/81547/222717717-4d5ed22d-d8a9-4b57-a022-cc7b9782ca63.png)

**Comment by grobinson-grafana** [comment]
Perhaps we can do both, and pick the larger of the two. If the time range is too small then the graph won't show enough information, while the same is also the case if the `For` interval is either 0 or just a couple minutes.

**Comment by apollo13** [comment]
Picking the larger one sounds like a good compromise. I am a bit afraid though that it might make even more sense to let the user actually define their own range because it might make sense to show more data in the graph to see trends easier which might not be clearly visible if just the evaluation range is considered. Nevertheless, your suggestion would already be a huge improvement and then time will tell if it is enough I guess :)

**Comment by grobinson-grafana** [comment]
Hi! I've opened a pull request to revert the default from 1h to 6h while I work on optimizing the date range based on a combination of `For` interval and the query range.

**Comment by sibebleuze** [comment]
Looking at the pull request mentioned in the initial comment (https://github.com/grafana/grafana/pull/59166), it seems to me like there were screenshot options added rather than a time range of 1 hour hardcoded. As these options are not exposed, the defaults are used every time, which is what the issue is about.
Would it be a good idea to expose these options in the alert configuration, in much the same way as is now done for the Dashboard UID and Panel ID?

If I'm not mistaken, these options are transferred from the alert configuration to the screenshot configuration in pkg/services/ngalert/image/service.go in the following block of code.
```
dashboardUID := r.GetDashboardUID()
if dashboardUID == "" {
  logger.Debug("Cannot take screenshot for alert rule as it is not associated with a dashboard")
  return nil, models.ErrNoDashboard
}

panelID := r.GetPanelID()
if panelID <= 0 {
  logger.Debug("Cannot take screenshot for alert rule as it is not associated with a panel")
  return nil, models.ErrNoPanel
}

logger = logger.New("dashboard", dashboardUID, "panel", panelID)

opts := screenshot.ScreenshotOptions{
  OrgID:        r.OrgID,
  DashboardUID: dashboardUID,
  PanelID:      panelID,
  Timeout:      s.screenshotTimeout,
}
```
Adding the time range options here would mean expanding the models.AlertRule struct (which is referenced as `r` in the code block) with a From and To field, along with some work to integrate them in the user interface.

As these options are quite specific to the Image Renderer plugin, it could be argued they don't belong in the general alert settings. I'm open to discussing other places where they would be reachable by the user, but I do think it would be valuable to be able to set these for each alert.

**Comment by BertSchel** [comment]
A simple workaround that works for now to show the panel in the time range you want it in your alert-messages:
Overrule the timepicker of the dashboard in your panel itself:
Query > Query options > Relative time > Set time to your likings.
This works for me.

**Comment by diborodin** [comment]
Hi there, 

I'm experiencing the same issue after upgrade from 9.3 to 9.5. I already tested it on 10.0.3 and this is still not fixed. This breaks my UX since I have had dozen of alerts where I need to observe a 2days/weekly/monthly periods in notification, presently the 1h-image in contact point looks absolutely pointless and confuses the users.

@grobinson-grafana do you think there is a chance to fix this in near future? I'd avoid downgrading to 9.3.

Thank you!

**Comment by BertSchel** [comment]
@diborodin For now, as a simple workaround I overrule the 1h-image by changing the relative time to my needs.
<img width="928" alt="overruleRelativeTime" src="https://github.com/grafana/grafana/assets/40671217/fd508c73-b3f9-4b3e-b525-47cce2e70793">
The image that goes with the alert will now show 7 days instead of 1h in this example.

**Comment by diborodin** [comment]
@BertSchel that really works, but it's so not obvious as there are no such settings in the rule creation interface. Anyway, I do appreciate your help! Thanks a lot!

**Comment by grobinson-grafana** [comment]
Something I thought about this afternoon is if we should set `From` and `To` based on the For duration of the alert rule. I put together a draft where:

1. The smallest possible screenshot (i.e. when For is 0 seconds) is the last 5 minutes
2. The largest possible screenshot is the last 24 hours
3. If the For duration is between 5 minutes and 60 minutes, we screenshot (2 * For) minutes. For example, if For is 5 minutes then the screenshot shows the last 10 minutes. If For is 30 minutes then the screenshot shows the last 1 hour.
4. If the For duration is more than 1 hour we screenshot (1 * For) minutes. For example, if For is 2 hours, the screenshot shows the last 2 hours, up to a maximum of 24 hours.

This is supposed to alleviate the issue where for example the default view of a dashboard might be the last 6 hours, but the data that is important for the alert happened in the last 5 minutes. However, because the screenshot contains the last 6 hours of data it's impossible to see what happened in the last 5 minutes because it's so small on the graph.

Would are your opinions on this?

**Comment by apollo13** [comment]
Wouldn't help us l. We really would want to show the whole panel range. I understand that this might not we the best default but depending on the calculations involved, the only sensible way is to be able to override the range from the alert config

**Comment by grobinson-grafana** [comment]
OK! Let's just use the default for the panel for the time being and then we can think about options later!

**Comment by apollo13** [comment]
That sounds like a good approach, thank you so much. Do you happen to know what prompted the initial changes? Ie was it based on user feedback?

**Comment by grobinson-grafana** [comment]
The initial change added options to allow customization of the time range when requesting screenshots, but a default of `1h` was applied when no customization was provided.

**Comment by grobinson-grafana** [comment]
This will be fixed in Grafana 10.2 🙂
